src/adapters/uniswap-v3/products/pool/uniswapV3PoolAdapter.ts

```diff
@@ -0,0 +1,354 @@
+import { getTokenMetadata } from '../../../../core/utils/getTokenMetadata'
+import {
+  ProtocolAdapterParams,
+  ProtocolDetails,
+  PositionType,
+  GetPositionsInput,
+  GetEventsInput,
+  MovementsByBlock,
+  GetTotalValueLockedInput,
+  GetProfitsInput,
+  GetApyInput,
+  GetAprInput,
+  GetClaimableRewardsInput,
+  GetConversionRateInput,
+  ProtocolRewardPosition,
+  ProtocolTokenApr,
+  ProtocolTokenApy,
+  ProtocolTokenUnderlyingRate,
+  ProfitsWithRange,
+  ProtocolTokenTvl,
+  ProtocolPosition,
+  TokenType,
+  Underlying,
+} from '../../../../types/adapter'
+import { Erc20Metadata } from '../../../../types/erc20Metadata'
+import { IProtocolAdapter } from '../../../../types/IProtocolAdapter'
+import { Protocol } from '../../../protocols'
+import { PositionManager__factory } from '../../contracts'
+
+const deadline = Math.floor(Date.now() - 1000) + 60 * 10
+
+const positionManagerCommonAddress =
+  '0xC36442b4a4522E871399CD717aBDD847Ab11FE88'
+
+const contractAddresses: Partial<Record<Chain, { positionManager: string }>> = {
+  [Chain.Ethereum]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Arbitrum]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Optimism]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Polygon]: {
+    positionManager: positionManagerCommonAddress,
+  },
+  [Chain.Bsc]: {
+    positionManager: '0x7b8A01B39D58278b5DE7e48c8449c9f4F5170613',
+  },
+  [Chain.Base]: {
+    positionManager: '0x03a520b32C04BF3bEEf7BEb72E919cf822Ed34f1',
+  },
+}
+
+const maxUint128 = BigInt(2) ** BigInt(128) - BigInt(1)
+
+export class UniswapV3PoolAdapter
+  implements IProtocolAdapter, IMetadataBuilder
+{
+  product = 'pool'
+  protocolId: Protocol
+  chainId: Chain
+
+  private provider: ethers.JsonRpcProvider
+
+  constructor({ provider, chainId, protocolId }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'UniswapV3',
+      description: 'UniswapV3 defi adapter',
+      siteUrl: 'https://uniswap.org/',
+      iconUrl:
+        'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png',
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+    }
+  }
+
+  /**
+   * Update me.
+   * Add logic to build protocol token metadata
+   * For context see dashboard example ./dashboard.png
+   * We need protocol token names, decimals, and also linked underlying tokens
+   */
+  @CacheToFile({ fileKey: 'protocol-token' })
+  async buildMetadata() {
+    throw new Error('Implement me')
+
+    return {}
+  }
+
+  /**
+   * Update me.
+   * Returning an array of your protocol tokens.
+   */
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
```

Comment:
We might want to reconsider whether this should be part of the `IProtocolAdapter` public interface.